hrms/hr/doctype/leave_allocation/leave_allocation.py

```diff
@@ -477,3 +442,86 @@ def get_unused_leaves(employee, leave_type, from_date, to_date):
 def validate_carry_forward(leave_type):
 	if not frappe.db.get_value("Leave Type", leave_type, "is_carry_forward"):
 		frappe.throw(_("Leave Type {0} cannot be carry-forwarded").format(leave_type))
+
+
+@frappe.whitelist()
+def bulk_allocate_leaves(allocations, new_leaves):
+	allocations = frappe.parse_json(allocations)
+	if len(allocations) <= 30:
+		return _bulk_allocate_leaves(allocations, new_leaves)
+
+	frappe.enqueue(_bulk_allocate_leaves, timeout=3000, allocations=allocations, new_leaves=new_leaves)
+	frappe.msgprint(
+		_("Allocation of leaves has been queued. It may take a few minutes."),
+		alert=True,
+		indicator="blue",
+	)
+
+
+def _bulk_allocate_leaves(allocations, new_leaves):
+	success, failure = [], []
+
+	for allocation in allocations:
+		doc = frappe.get_doc("Leave Allocation", allocation)
+		try:
+			result = process_leave_allocation(doc, new_leaves)
+			success.append(result)
+		except Exception as e:
+			frappe.log_error(
+				title=f"Bulk Allocation - Processing failed for Allocation {doc.name}.",
+				message=str(e),
+				reference_doctype="Leave Allocation",
+				reference_name=doc.name,
+			)
+			failure.append(doc.employee)
+
+	frappe.clear_messages()
+	frappe.publish_realtime(
+		"completed_bulk_leave_allocation",
+		message={"success": success, "failure": failure, "for_update": True},
+		doctype="Leave Allocation",
+		after_commit=True,
+	)
+
+
+def process_leave_allocation(doc, new_leaves, date=None):
```

Comment:
This can go inside the class?

If you need to pass a doc to a function, then it's better to encapsulate it within the class. 
`doc.process_leave_allocation`
better than
`process_leave_allocation(doc)`

Reply:
Also instead of process_leave_allocation, should we call it `doc.add_leaves`?
That's more descriptive